Add transaction processor support to the Rust SDK

vaporos opened the pull request.

peterschwarz commented on sdk/rust/src/processor/handler.rs:
```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright 2017 Bitwise IO, Inc.
```
2018?

vaporos replied:
2017 is accurate, it started late last year.

peterschwarz commented on sdk/rust/src/processor/mod.rs:
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 Bitwise IO, Inc.
```
2018?

peterschwarz commented on sdk/rust/src/processor/mod.rs:
```diff
@@ -0,0 +1,144 @@
+                                x).unwrap();
+                        },
+                        _ => {
+                            let mut response = TpProcessResponse::new();
```
I'm not sure what sort of behavior this will result in: sending back a response is good (i.e. the future on the other end will resolve), but it will potentially be the wrong type of message.

peterschwarz replied:
Missing handling of `PING_REQUEST`, thought this could probably be done in a future PR

agunde406 replied:
I can include the Ping functionality with the reconnect work.

vaporos replied:
We could also include the default handling rework in with reconnect work (probably some affinity to testing it all together).

aludvik commented on sdk/rust/src/processor/handler.rs:
```diff
@@ -0,0 +1,243 @@
+        match *self {
+            ContextError::AuthorizationError(ref s) =>
+                write!(f, "AuthorizationError: {}", s),
+            ContextError::UnknownError(ref s) =>
```
UnknownError seems like it could be abused rather easily. It looks like this is just being used when validating a response from the validator, so maybe something like `ConsistencyError` or `IntegrityError`?

aludvik commented on sdk/rust/src/processor/handler.rs:
```diff
@@ -0,0 +1,243 @@
+        let serialized = request.write_to_bytes()?;
+        let x : &[u8] = &serialized;
+
+        let correlation_id: String = rand::thread_rng().gen_ascii_chars().take(16).collect();
```
In other implementations, correlation id generation is behind another function, which is what I would expect here. If not, at a minimum 16 should be moved to a constant.

vaporos replied:
This is now a function generate_correlation_id().

aludvik commented on sdk/rust/src/processor/mod.rs:
```diff
@@ -0,0 +1,144 @@
+                    x).unwrap();
+
+                // Absorb the TpRegisterResponse message
+                let _ = future.get().unwrap();
```
Why are we unwrapping here?

aludvik commented on sdk/rust/src/processor/mod.rs:
```diff
@@ -0,0 +1,144 @@
+        loop {
+            match receiver.recv() {
+                Ok(r) => {
+                    let message = r.unwrap();
```
Why are we unrwapping?

aludvik commented on sdk/rust/src/processor/mod.rs:
```diff
@@ -0,0 +1,144 @@
+                                }
+                            };
+
+                            let serialized = response.write_to_bytes().unwrap();
```
Why wrapping unwrapping?

aludvik commented on sdk/rust/src/processor/mod.rs:
```diff
@@ -0,0 +1,144 @@
+                            sender.reply(
+                                Message_MessageType::TP_PROCESS_RESPONSE,
+                                message.get_correlation_id(),
+                                x).unwrap();
```
Why are we unrwapping?

vaporos commented:
Rebased and updated "Add transaction processor support to Rust SDK" to not add rand to Cargo.toml, as that was added in master recently as part of the sawadm PR.

vaporos commented:
With respect to the error handling deficiencies in start() (the unwrap()s), I think we need to reconne-ction logic to properly handle many of these.  Or at least it seems related from a flow perspective.

For example, if we receive an error prior to the current loop(), such as during registration, we may want to disconnect and try again.  Because this could happen on re-connection, we shouldn't return an error from start() as we don't want transaction processors to exit due to transient errors.
